src/check/parse/tokenize/src/main.zig

```diff
@@ -0,0 +1,1425 @@
+        }
+    }
+
+    pub fn chompIdentUpper(self: *Cursor) T {
+        while (self.pos < self.buf.len) {
+            const c = self.buf[self.pos];
+            if ((c >= 'a' and c <= 'z') or (c >= 'A' and c <= 'Z') or (c >= '0' and c <= '9')) {
+                self.pos += 1;
+            } else {
+                break;
+            }
+        }
+        return T.UpperIdent;
+    }
+
+    pub fn chompInteger(self: *Cursor) void {
+        while (self.pos < self.buf.len) {
+            const c = self.buf[self.pos];
+            if (c >= '0' and c <= '9') {
+                self.pos += 1;
+            } else {
+                break;
+            }
+        }
+    }
+
+    pub fn chompIdentGeneral(self: *Cursor) void {
+        while (self.pos < self.buf.len) {
+            const c = self.buf[self.pos];
+            if ((c >= 'a' and c <= 'z') or (c >= 'A' and c <= 'Z') or (c >= '0' and c <= '9') or c == '_') {
+                self.pos += 1;
+            } else {
+                const info = self.peekUnicode();
+                if (info.kind != UnicodeKind.Other and info.kind != UnicodeKind.Invalid) {
+                    self.pos += info.length;
+                } else {
+                    break;
+                }
+            }
+        }
+    }
+};
+
+/// The output of the tokenizer.
+pub const TokenOutput = struct {
+    tokens: TokenizedBuffer,
+    /// The messages slice is the same one provided by the caller.
+    messages: []Message,
+    message_count: usize,
+};
+
+const BraceKind = enum {
+    Round,
+    Square,
+    Curly,
+    StringInterpolation,
+    StringInterpolationMultiline,
+    SingleQuoteInterpolation,
+    SingleQuoteInterpolationMultiline,
+};
+
+/// The tokenizer that uses a Cursor and produces a TokenizedBuffer.
+pub const Tokenizer = struct {
+    cursor: Cursor,
+    output: TokenizedBuffer,
+    stack: std.ArrayList(BraceKind),
+
+    /// Creates a new Tokenizer.
+    /// Note that the caller must also provide a pre-allocated messages buffer.
+    pub fn init(text: []const u8, messages: []Message, gc: *GenCatData, allocator: std.mem.Allocator) !Tokenizer {
+        const cursor = Cursor.init(text, messages, gc);
+        var output = try TokenizedBuffer.init(allocator);
+        // Push an initial line with indent 0.
+        try output.lines.append(Line{
+            .token_start = 0,
+            .indent = Indent.init(),
+        });
+        return Tokenizer{
+            .cursor = cursor,
+            .output = output,
+            .stack = std.ArrayList(BraceKind).init(allocator),
+        };
+    }
+
+    pub fn destroy(self: *Tokenizer) void {
+        // self.output.kinds.deinit();
+        // self.output.offsets.deinit();
+        // self.output.lengths.deinit();
+        // self.output.lines.deinit();
+        self.stack.deinit();
+        self.gc.deinit();
+    }
+
+    // A simplified equivalent to the Rust macros:
+    inline fn pushToken(self: *Tokenizer, kind: T, start: usize) !void {
```

Comment:
recommend to abstain from `inline fn` unless there is a specific reason for it:

[from the docs](https://ziglang.org/documentation/master/#inline-fn):

> It is generally better to let the compiler decide when to inline a function, except for these scenarios:
> * To change how many stack frames are in the call stack, for debugging purposes.
> * To force comptime-ness of the arguments to propagate to the return value of the function, as in the above example.
> * Real world performance measurements demand it.
> Note that inline actually restricts what the compiler is allowed to do. This can harm binary size, compilation speed, and even runtime performance.